### PR TITLE
Reconcile when rollout duration is changed

### LINKED ISSUE #2946: Fail to reconcile rollout duration when isvc updated
/kind bug

**What steps did you take and what happened:**
[A clear and concise description of what the bug is.]

- Deploy a isvc without rollout duration annotation.
- Update isvc with rollout duration annotation .
- ksvc rollout duration is not unchanged.

**What did you expect to happen:**
ksvc rollout duration is properly updated

**What's the InferenceService yaml:**
[To help us debug please run `kubectl get isvc $name -n $namespace -oyaml` and paste the output]

Deploy

```yaml
apiVersion: serving.kserve.io/v1beta1
kind: InferenceService
metadata:
  name: flower-sample
spec:
  predictor:
    model:
      modelFormat:
        name: tensorflow
      storageUri: "gs://kfserving-examples/models/tensorflow/flowers"
```

and update with

```yaml
apiVersion: serving.kserve.io/v1beta1
kind: InferenceService
metadata:
  name: flower-sample
  annotations:
    serving.knative.dev/rollout-duration: "30s"
spec:
  predictor:
    model:
      modelFormat:
        name: tensorflow
      storageUri: "gs://kfserving-examples/models/tensorflow/flowers"
```

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- Istio Version:
- Knative Version:
- KServe Version: v0.10.0
- Kubeflow version:
- Cloud Environment:[k8s_istio/istio_dex/gcp_basic_auth/gcp_iap/aws/aws_cognito/ibm]
- Minikube/Kind version:
- Kubernetes version: (use `kubectl version`):
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by henrysecond1**
**What this PR does / why we need it**:

Reconcile when rollout duration is changed.
Currently, ksvc reconciler does not reconcile ksvc when rollout duration is changed

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #2946 

**Type of changes**
Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**Feature/Issue validation/testing**:

Provide instructions so it can be reproduced.

- Deploy a isvc without rollout duration annotation.
- Update isvc with rollout duration annotation .
- Confirm that ksvc rollout duration remains unchanged.

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Checklist**:

- [ ] Have you added unit/e2e tests that prove your fix is effective or that this feature works?
- [ ] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

**Release note**:

```release-note
none
```

### CODE REVIEW

File: pkg/controller/v1beta1/inferenceservice/reconcilers/knative/ksvc_reconciler.go [hunk 1]
```diff
@@ -209,6 +209,9 @@ func (r *KsvcReconciler) Reconcile() (*knservingv1.ServiceStatus, error) {
 	existing.Spec.ConfigurationSpec = desired.Spec.ConfigurationSpec
 	existing.ObjectMeta.Labels = desired.ObjectMeta.Labels
 	existing.Spec.Traffic = desired.Spec.Traffic
+	if value, ok := desired.ObjectMeta.Annotations[constants.RollOutDurationAnnotationKey]; ok {
```
Comment (yuzisun) [inline pkg/controller/v1beta1/inferenceservice/reconcilers/knative/ksvc_reconciler.go]: @henrysecond1 Is this the only field at the metadata annotation level?

Comment (yuzisun) [reply pkg/controller/v1beta1/inferenceservice/reconcilers/knative/ksvc_reconciler.go]: We might want to implement it in a more generic way to reconcile the annotations under knative serving api group, so we do not need handle this one off case.

**Comment by yuzisun** [comment]
@henrysecond1 Can you help create an bug issue and then link to this PR? thanks

File: pkg/controller/v1beta1/inferenceservice/reconcilers/knative/ksvc_reconciler.go (hunk 1 above)
Comment (henrysecond1) [reply pkg/controller/v1beta1/inferenceservice/reconcilers/knative/ksvc_reconciler.go]: > @henrysecond1 Is this the only field at the metadata annotation level?

@yuzisun AFAIK, yes it's the only field at ksvc metadata annotation level (https://github.com/kserve/kserve/blob/v0.10.1/pkg/controller/v1beta1/inferenceservice/reconcilers/knative/ksvc_reconciler.go#L97-L103)

> We might want to implement it in a more generic way to reconcile the annotations under knative serving api group, so we do not need handle this one off case.

I understand, but when I tried to reconcile whole annotations, knative does not allow it with validation errors `"validation.webhook.serving.knative.dev" denied the request: validation failed: annotation value is immutable: metadata.annotations.serving.knative.dev/creator`

or we can maintain set of annotations managed by kserve like

```go
var managedKsvcAnnotations = map[string]bool{
	constants.RollOutDurationAnnotationKey: true,
}

...

	for annotationKey := range managedKsvcAnnotations {
		if value, ok := desired.ObjectMeta.Annotations[annotationKey]; ok {
			existing.ObjectMeta.Annotations[annotationKey] = value
		}
	}
```

I think it's a more generic way. what do you think about it?

**Comment by henrysecond1** [comment]
@yuzisun Thanks for taking a look. I created a bug issue and linked it.

File: pkg/controller/v1beta1/inferenceservice/reconcilers/knative/ksvc_reconciler.go (hunk 1 above)
Comment (yuzisun) [reply pkg/controller/v1beta1/inferenceservice/reconcilers/knative/ksvc_reconciler.go]: @henrysecond1 sounds good, we need to reconcile this explicit list of ksvc annotations.

Comment (henrysecond1) [reply pkg/controller/v1beta1/inferenceservice/reconcilers/knative/ksvc_reconciler.go]: @yuzisun added a new commit for it. PTAL

File: pkg/controller/v1beta1/inferenceservice/reconcilers/knative/ksvc_reconciler.go [hunk 2]
```diff
@@ -220,6 +231,13 @@ func (r *KsvcReconciler) Reconcile() (*knservingv1.ServiceStatus, error) {
 }
 
 func semanticEquals(desiredService, service *knservingv1.Service) bool {
+	for ksvcAnnotationKey := range managedKsvcAnnotations {
+		if desiredValue, ok := desiredService.ObjectMeta.Annotations[ksvcAnnotationKey]; ok {
+			if service.ObjectMeta.Annotations[ksvcAnnotationKey] != desiredValue {
```
Comment (yuzisun) [inline pkg/controller/v1beta1/inferenceservice/reconcilers/knative/ksvc_reconciler.go]: I think we need additional check here as the annotation may not exist on the current service or it exists but you removed it from desired.

Comment (henrysecond1) [reply pkg/controller/v1beta1/inferenceservice/reconcilers/knative/ksvc_reconciler.go]: @yuzisun that's correct. I modified the function and tested it with several cases. I think all cases work as expected

- add new annotation to isvc -> annotation is added to ksvc
- modify existing annotation in isvc -> annotation is modified in ksvc
- remove existing annotation from isvc -> annotation is removed in ksvc
- manually add / modify / remove annotation in ksvc -> reconciled

**Comment by yuzisun** [comment]
Thanks @henrysecond1 !
/lgtm
/approve

**Pull request merged**